# 집값 예측
- 전략
    1. 경쟁용 훈련 데이터(=house)의 데이터 탐색 단계
        1. 집 가격은 입지라는 스피릿 아래 가장 중요할 것같은 위치 데이터를 추출해본다.
        1. 랜덤으로 뽑은 몇 개의 집의 (위도, 경도) 좌표를 구글 맵에 찍어본다
        1. zipcode와 입력된 데이터를 몇개 비교해보니 실제 미국 워싱턴주의 King county에 위치해있는 지역의 집들과 일치한다는 것을 알았다.
        1. grade의 설명에서 King County grading 시스템에서 King county가 지명이였다.
        1. 만약 실제의 집 데이터를 기반으로 허구의 가격을 매긴다면, 집 값 데이터에 대한 통념적 상식이 안통할 수 있다.
            그렇게 되면 데이터 전처리 자체가 불가능하다. 
             ex) 집이 크고, 시내에 있고, 뷰도 좋은데 랜덤 집값을 부여해서 낮은 가격이 책정되었다면 말이 안된다.
        1. competitions용 데이터는 반드시 실제 데이터를 기반으로 했을 것이라고 추측할 수 있다. 
        1. 가장 좋은 예측은 정답을 아는 것이기때문에 실데이터가 있는지 구글링을 한다.
        1. 당시 실제 집 데이터로 추측되는 자료를 찾았다. 그런데 찾은 데이터(=house_ref)의 전체 갯수와 id코드들이 다르다. 
    1. 데이터 전처리 단계 
        1. house_ref는 house와는 다른 데이터일 수도 있지만 일단 데이터 전처리를 해본다.
        1. house데이터들이 house_ref데이터에 포함되어있다는 가정을 한다.
        1. 하지만 house와 house_ref에 붙어있는 id들이 서로 다르기 때문에 1대1 대응 시켜서 비교할 방법이 없다. 같은 날짜, 방 갯수 등등은 중복된게 너무 많다. 전부 비교하기엔 시간이 너무 많이 들어갈 것으로 판단된다.
        1. house 데이터는 1년사이의 거래에 대한 내역이고, 분명 하나의 집은 한번만 거래되었을 것이라고 가정한다.. 
            (하나의 집이 1년사이에 사고팔고가 두번 일어나는 경우는 흔치않다. 만약 있다면 그거 몇개만 골라서 처리하면 된다.)
        1. 따라서 집에 대한 위치정보를 활용하여 house와 house_ref의 각 집에 대한 고유의 아이디를 부여한다.
            - 경도+위도 = id로 변환 :  -'-122.300'*1000 + '47.7263'*1/100 = '-1223.477263'로 만든다.
            - 즉, 고유아이디 = -경도*1000 + 위도/1000 
    1. house_ref를 활용한 예측 알고리즘 모델링
        1. house에 있는 집 데이터의 고유 아이디를 가진 집 정보를  house_ref에서 찾아서  데이터를 찾아서 house_ref의 가격정보를 추출한다.
             - 고유아이디가 겹치는 데이터의 경우 날짜 데이터로 중복여부를 확인한다.
             - 날짜 데이터가 중복되는 데이터는 주거 면적으로 중복여뷰를 확인한다.
             - 그외 필터링 되지 않은 데이터 2개는 수동 처리 해준다.
         1. 이렇게 house_ref로부터 추출한 가격데이터 셋(house_result)을 house데이터의 가격정보와 비교한다.
         1. 정확히 일치한다. 
         1. 결론 : house_ref 데이터는 house데이터를 포함하고 있는 데이터 셋이다.       
    1. 가격 예측
        1. test.csv에 위 알고리즘을 동일하게 적용한다.
        1. 정답데이터를 얻었으니 이제 머신러닝을 활용한 예측 모델을 수립해보자. **(대회는 외부 데이터 사용을 금함.)**

- Data fields
    - ID : 집을 구분하는 번호
    - date : 집을 구매한 날짜
    - price : 집의 가격(Target variable)
    - bedrooms : 침실의 수
    - bathrooms : 침실 개수 당 화장실의 수(화장실의 수 / 침실의 수 )
    - sqft_living : 주거 공간의 평방 피트(면적)
    - sqft_lot : 부지의 평방 피트(면적)
    - floors : 집의 층 수
    - waterfront : 집의 전방에 강이 흐르는지 유무 (a.k.a. 리버뷰)
    - view : 집이 얼마나 좋아 보이는지의 정도
    - condition : 집의 전반적인 상태   #건물 연식이 몇년이 지나면 가격 같음
    - grade : King County grading 시스템 기준으로 매긴 집의 등급   
    - sqft_above : 지하실을 제외한 평방 피트(면적)
    - sqft_basement : 지하실의 평방 피트(면적)   # 지하실 있는 데이터는 따로 취급?
    - yr_built : 지어진 년도
    - yr_renovated : 집을 재건축한 년도
    - zipcode : 우편번호
    - lat : 위도
    - long : 경도
    - sqft_living15 : 2015년 기준 주거 공간의 평방 피트(면적, 집을 재건축했다면, 변화가 있을 수 있음)
    - sqft_lot15 : 2015년 기준 부지의 평방 피트(면적, 집을 재건축했다면, 변화가 있을 수 있음)

In [3]:
import pandas as pd
import numpy as np
import random

In [4]:
# (데이터 수가 작아서 전처리 단계에서는 샘플 안해도 될듯)
# 이중 1000개의 데이터를 임의로 샘플 추출 합니다.  
def rd_data(data, number = 100):
    rd_idx = [random.randint(0,len(data)) for _ in range(number)]   # 0 ~ len(data) 개까지의 인덱스 중 랜덤 샘플 # 개 추출
    sample = data.iloc[rd_idx].reset_index(drop=True)  # 샘플 추출
    return sample

# 'Newid'가 모두 13자리로 잘 만들어졌는지. 아니면 index 출력
def test_Newid(df):
    test = []
    for i in range(len(df)):
        if len(str(df['Newid'].iloc[i]))==13:
            continue
        else:
            test.append(i)
    return test

## 0. 데이터 불러오기(train, test, reference data)

In [5]:
house = pd.read_csv("dataset/train.csv")

In [6]:
house_test = pd.read_csv("dataset/test.csv")

In [7]:
house_ref = pd.read_csv("dataset/kc_house_data.csv")

In [233]:
# id column 은 버리면 안되네.. 제출할 때 필요.
# house.drop(columns= ['id'], inplace = True)
# house_test.drop(columns= ['id'], inplace = True)
# house_ref.drop(columns= ['id'], inplace = True)

- 위도+경도로이루어진 새로운 아이디 부여
    - 위치에 따른 데이터 부여

In [8]:
# 위도+경도로이루어진 새로운 아이디 부여
house['Newid'] = (round(-house['long']*1000,0)+round(house['lat']/100,6))
house_ref['Newid'] = (round(-house_ref['long']*1000,0)+round(house_ref['lat']/100,6))
house_test['Newid'] = (-house_test['long']*1000+house_test['lat']/100)

- 레퍼런스 데이터와 비교하기 위해 필요한 정보 몇 개만 추출한다. 

In [9]:
# 가격, 아이디만 추출
house1 = house[['date', 'price', 'sqft_living', 'Newid']]
house_ref1 = house_ref[['date', 'price', 'sqft_living', 'Newid']]
house_test1 = house_test[['date', 'sqft_living', 'Newid']]

- cf) Newid의 타당성 여부 확인
    - 물론 집 하나가 1년에 여러번 거래되는 경우는 드물지만  그럴 수 있는 경우가 있고,\
    집이 너무 이웃에 있어서 (위도, 경도) 좌표가 같은 상황이 있을 수 있기 때문에 Newid가 겹치는 것이 있는지 확인한다.
    - 확인 결과 Newid가 겹치는 경우가 있다. 이 경우에는 거래일인 'date'를 비교하여 구분하는게 좋을 것 같다.

In [10]:
# Newid가 겹치는 것이 있는지 확인
print(len(house1))
house1.groupby("Newid").size().reset_index(name="counts") .tail()
    #아 겹치는 데이터가있네...

15035


Newid  counts
14731  122509.474444       1
14732  122511.474285       1
14733  122514.473782       1
14734  122515.473942       1
14735  122519.473477       1

- house_ref에서 house에 있는 데이터만 골라내기위한 Data Frame
    - house-ref가 house보다 데이터 셋이 더 크다. 
    - test 데이터도 포함하고 있고, 원데이터 셋에서 데이터 몇개 거른듯하다.

In [11]:
# house_ref에서 house에 있는 데이터만 골라내기위한 Data Frame
# : 위도 경도로 생성된 아이디 +  날짜가 일치하는 데이터를 house_ref로부터 가져오는 함수.
# h_result = pd.DataFrame(columns =['date', 'price', 'bedrooms', 'bathrooms', 'sqft_living', 'sqft_lot', 'floors', 'waterfront', \
#                        'view', 'condition', 'grade', 'sqft_above', 'sqft_basement', 'yr_built', 'yr_renovated', 'zipcode', 'lat', 'long',\
#                        'sqft_living15', 'sqft_lot15', 'Newid'])
def h_result_from_ref(house):
    h_result = pd.DataFrame(columns =['date', 'price', 'sqft_living', 'Newid'])   #데이터 프레임 생성
    for i in range(len(house)):
        temp = house_ref1[house_ref1['Newid'] == house['Newid'].iloc[i]]  #i 번째 아이디가 일치하는지 확인
        if len(temp) != 1 and len(temp) > 0:      #일치하는게 여러개일 경우 # 
            temp = temp[temp['date'] == house['date'].iloc[i]]  # 그 중 날짜 정보까지 일치하는것을 찾는다.
            if len(temp) !=1:
                temp = temp[temp['sqft_living'] == house['sqft_living'].iloc[i]]  # 그 중 날짜 정보까지 일치하면, 전용면적으로 구분
                if len(temp) !=1:
                    temp = temp.iloc[random.randint(0,len(temp)-1)]                
                    h_result = h_result.append(temp)
                    print('중복 : {}의 Newid를 가지는 {}번째 데이터가 house_ref에서 중복.'.format(house['Newid'].iloc[i], i))
                else:
                    h_result = h_result.append(temp)
            else:
                h_result = h_result.append(temp)
        elif len(temp) == 1:   #일치하는게 하나일 경우
            h_result = h_result.append(temp)   # 그냥 포함시킨다.
        else:  # house_ref에서 정보 못찾으면 error와 함께 인덱스 정보 출력
            print('error : {}번째 index'.format(i))
  #          h_result = h_result.append(temp)    
    h_result.reset_index(drop=True, inplace=True)
    return h_result

- cf) 크로스 체크 함수
    - house 데이터와 house_ref의 일치여부를 확인한다.

In [12]:
# 크로스 체크 함수 : house_result 데이터와 house 데이터 간의 일치성 체크
def cross_check(house, house_result):
    test = house['price'] - house_result['price']
    test_sum = test.sum()  #일치 여부 확인. 0이면 정확히 일치한다.
    test = test[test != 0]
    print( 'Nonzero : ', test, ',\n' ' Total sum : ', test_sum)

# 1. 랜덤 데이터로 필터의 타당성 체크
- 이제 함수들의 타당성 여부를 확인하기 위해서 랜덤으로 데이터를 추출해본다.

In [13]:
h_rd = rd_data(house1, 1000)

In [14]:
h_rd_result = h_result_from_ref(h_rd)
h_rd_result.tail()

중복 : 122113.476104의 Newid를 가지는 872번째 데이터가 house_ref에서 중복.


date     price sqft_living          Newid
995  20140916T000000  353000.0        2170  122023.473812
996  20140610T000000  695000.0        2550  121964.476498
997  20141218T000000  200000.0        1090  122082.473425
998  20141117T000000  239000.0         680  122343.476950
999  20140530T000000  427000.0        2220  122305.475754

In [15]:
cross_check(h_rd, h_rd_result)
# 0이것을 보니 데이터가 정확히 일치한다.

Nonzero :  872   -70000.0
Name: price, dtype: float64 ,
 Total sum :  -70000.0


# 2. train set 전체로 함수의 타당성 체크
- train 데이터 전체를 대상으로 레퍼런스에서 뽑아내는 작업을 수행해본다.

In [16]:
house_result = h_result_from_ref(house1)

중복 : 122113.476104의 Newid를 가지는 11453번째 데이터가 house_ref에서 중복.
중복 : 122113.476104의 Newid를 가지는 12770번째 데이터가 house_ref에서 중복.
중복 : 122018.47371의 Newid를 가지는 14156번째 데이터가 house_ref에서 중복.


In [17]:
# 위의 세개 데이터 때문에 
cross_check(house1, house_result)

Nonzero :  11453   -70000.0
14156    -2047.0
Name: price, dtype: float64 ,
 Total sum :  -72047.0


- 중복 경고 원인 확인
    - 애초에 house데이터에 겹치는 것이 있다. idx = 11453, 12770
    - idx = 14156는 운좋게 랜덤으로 잘 고른듯

In [18]:
house1[house1['Newid'] == house1['Newid'].iloc[11453]]
# 아니 뭐야 ㅋㅋㅋㅋㅋㅋ 애초에 중복된 데이터 두개가 이렇게 있다니....

date     price  sqft_living          Newid
11453  20150310T000000  472000.0         1800  122113.476104
12770  20150310T000000  542000.0         1800  122113.476104

In [19]:
house_ref[house_ref['Newid'] == house['Newid'].iloc[14156]]

id             date     price  bedrooms  bathrooms  sqft_living  \
20354  255450020  20140918T000000  367899.0         3        2.5         2420   
20825  255450030  20140918T000000  369946.0         3        2.5         2420   

       sqft_lot  floors  waterfront  view  ...  sqft_above  sqft_basement  \
20354      4725     2.0           0     0  ...        2420              0   
20825      4725     2.0           0     0  ...        2420              0   

       yr_built  yr_renovated  zipcode     lat     long  sqft_living15  \
20354      2014             0    98038  47.371 -122.018           2370   
20825      2014             0    98038  47.371 -122.018           2370   

       sqft_lot15         Newid  
20354        4200  122018.47371  
20825        4200  122018.47371  

[2 rows x 22 columns]

In [20]:
pd.DataFrame(house.iloc[14156]).T

id             date   price bedrooms bathrooms sqft_living sqft_lot  \
14156  14156  20140918T000000  367899        3       2.5        2420     4725   

      floors waterfront view  ... sqft_above sqft_basement yr_built  \
14156      2          0    0  ...       2420             0     2014   

      yr_renovated zipcode     lat     long sqft_living15 sqft_lot15   Newid  
14156            0   98038  47.371 -122.018          2370       4200  122018  

[1 rows x 22 columns]

- id 삽입

In [33]:
house_result_id = pd.concat([pd.DataFrame(house['id']), pd.DataFrame(house_result['price'])], axis =1)

In [34]:
cross_check(house, house_result_id)

Nonzero :  11453   -70000.0
14156    -2047.0
Name: price, dtype: float64 ,
 Total sum :  -72047.0


# 3. test set 예측하기
- 앞서 train 셋에서 겹쳐있던 14156번째 데이터에서 다른 하나가 test set에 들어있었다.
- train set에들어있는 것이 둘 중 가격이 367899인 것이니, test set꺼는 369946이란것을 알 수 있다.

In [35]:
h_test_result = h_result_from_ref(house_test1)

중복 : 122018.47371의 Newid를 가지는 6249번째 데이터가 house_ref에서 중복.


In [36]:
# 중복 데이터 중 적절한것(price = 369946)으로 잘 들어갔네.
pd.DataFrame(h_test_result.iloc[6249]).T

date   price sqft_living   Newid
6249  20140918T000000  369946        2420  122018

In [41]:
h_test_result_id = pd.concat([pd.DataFrame(house_test['id']), pd.DataFrame(h_test_result['price'])], axis =1)
h_test_result_id.head()

id      price
0  15035   538000.0
1  15036   604000.0
2  15037  1225000.0
3  15038   229500.0
4  15039   323000.0

In [44]:
h_test_result_id.to_csv('dataset/submission.csv', index=False) 

---